cmd/nmi/main.go

```diff
@@ -114,7 +114,10 @@ func main() {
 	if err = metrics.RegisterAndExport(*prometheusPort); err != nil {
 		klog.Fatalf("Could not register and export metrics: %+v", err)
 	}
-	if err := s.Run(); err != nil {
+
+	rd := s.GetRedirector()
+
+	if err := s.Run(rd); err != nil {
 		klog.Fatalf("%s", err)
 	}
```

Comment:
When run completes main will complete too and NMI will exit immediately.
It does not wait for redirector goroutine to perform graceful shutdown - which may leave IP tables not being cleaned up on Linux and routing not cleaned up on Windows.
This looks like a preexisting bug which needs to be fixed.

Reply:
In redirector goroutine itself, it will try to wait for interrupt and terminal signal, then do the cleanup for ITTable and route policy. See handleTermination function in redirector_linux.go file. Do you think this will solve the problem here?

---
In reply to: [383552668](https://github.com/bowu24/aad-pod-identity/pull/1#discussion_r383552668) [](ancestors = 383552668)

Reply:
I do not think that waiting in go routine will address this issue. The moment main() exits all goroutines will be terminated unconditionally. to solve the problem you need to make main() to wait for an event from goroutine before exiting - this can be done via closing a channel

---
In reply to: [386542574](https://github.com/bowu24/aad-pod-identity/pull/1#discussion_r386542574) [](ancestors = 386542574,383552668)

Reply:
Got it. That makes sense. Just curious if I could I also use sync.WaitGroup to implement it though there may be only goroutine in this case?

---
In reply to: [386636021](https://github.com/bowu24/aad-pod-identity/pull/1#discussion_r386636021) [](ancestors = 386636021,386542574,383552668)

Reply:
Update the code. Let me know if you have any comments.